File: packages/gavio-js/tests/unit/erasure.test.ts
import { describe, it, expect } from 'vitest'
import { mkdtempSync, readFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { Gateway } from '../../src/gateway.js'
import { auditInterceptor, AuditRecord, jsonlSink } from '../../src/interceptors/audit/index.js'
import { mockProvider } from '../../src/providers/mock.js'

function rec(subjectId: string | null, traceId = 't'): AuditRecord {
  return new AuditRecord({ traceId, provider: 'mock', model: 'm', timestampUtc: 'now', subjectId })
}

function gw() {
  return new Gateway({ model: 'mock' }).withAdapter(mockProvider({ response: 'hi' })).use(auditInterceptor())
}

describe('right-to-erasure (F-QUA-09)', () => {
  it('persists subjectId from request metadata into the audit record', async () => {
    const res = await gw().complete({
      messages: [{ role: 'user', content: 'q' }],
      metadata: { subject_id: 'user-123' },
    })
    expect(res.audit?.subjectId).toBe('user-123')
  })

  it('leaves subjectId null when absent', async () => {
    const res = await gw().complete({ messages: [{ role: 'user', content: 'q' }] })
    expect(res.audit?.subjectId).toBeNull()
  })

  it('jsonlSink.purge removes matching records and returns the count', async () => {
    const path = join(mkdtempSync(join(tmpdir(), 'gavio-erasure-')), 'audit.jsonl')
    const sink = jsonlSink({ path })
    await sink.write(rec('u1', 't1'))
    await sink.write(rec('u2', 't2'))
    await sink.write(rec('u1', 't3'))

    const removed = await sink.purge!('u1')

    expect(removed).toBe(2)
    const lines = readFileSync(path, 'utf8')
      .split('\n')
      .filter(Boolean)
      .map((l) => JSON.parse(l))
    expect(lines).toHaveLength(1)
    expect(lines[0].subjectId).toBe('u2')
  })

  it('purge returns 0 on a missing file and on no match', async () => {
    const path = join(mkdtempSync(join(tmpdir(), 'gavio-erasure-')), 'audit.jsonl')
    const sink = jsonlSink({ path })
    expect(await sink.purge!('nobody')).toBe(0) // file not yet written
    await sink.write(rec('u1'))
    expect(await sink.purge!('nobody')).toBe(0)
  })
})

File: packages/gavio-java/gavio-interceptor-audit/src/test/java/io/gavio/interceptors/audit/ErasureTest.java
package io.gavio.interceptors.audit;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

import io.gavio.Gateway;
import io.gavio.GavioRequest;
import io.gavio.interceptors.audit.sinks.JsonlSink;
import io.gavio.json.Json;
import io.gavio.providers.MockProvider;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class ErasureTest {

    private static final class CollectingSink implements AuditSink {
        final List<AuditRecord> records = new ArrayList<>();

        @Override
        public CompletableFuture<Void> write(AuditRecord record) {
            records.add(record);
            return CompletableFuture.completedFuture(null);
        }
    }

    private static AuditRecord rec(String subjectId, String trace) {
        return AuditRecord.builder()
                .traceId(trace)
                .provider("mock")
                .model("m")
                .subjectId(subjectId)
                .build();
    }

    private static Gateway gw(AuditSink sink) {
        return Gateway.builder()
                .adapter(new MockProvider("hi"))
                .model("mock")
                .use(new AuditInterceptor(sink))
                .build();
    }

    @Test
    void subjectIdPersistedFromMetadata() {
        CollectingSink sink = new CollectingSink();
        GavioRequest req = GavioRequest.builder()
                .message("user", "q")
                .model("mock")
                .metadata("subject_id", "user-123")
                .build();
        gw(sink).complete(req).join();
        assertEquals("user-123", sink.records.get(0).subjectId());
    }

    @Test
    void subjectIdNullWhenAbsent() {
        CollectingSink sink = new CollectingSink();
        gw(sink).complete(GavioRequest.builder().message("user", "q").model("mock").build()).join();
        assertNull(sink.records.get(0).subjectId());
    }

    @Test
    void jsonlPurgeRemovesMatchingAndReturnsCount(@TempDir Path dir) throws Exception {
        Path path = dir.resolve("audit.jsonl");
        JsonlSink sink = new JsonlSink(path);
        sink.write(rec("u1", "t1")).join();
        sink.write(rec("u2", "t2")).join();
        sink.write(rec("u1", "t3")).join();

        int removed = sink.purge("u1").join();

        assertEquals(2, removed);
        List<String> lines =
                Files.readAllLines(path).stream().filter(l -> !l.isBlank()).toList();
        assertEquals(1, lines.size());
        assertEquals("u2", Json.parseObject(lines.get(0)).get("subject_id"));
    }

    @Test
    void purgeZeroOnMissingFileAndNoMatch(@TempDir Path dir) throws Exception {
        JsonlSink sink = new JsonlSink(dir.resolve("audit.jsonl"));
        assertEquals(0, sink.purge("nobody").join()); // file not yet written
        sink.write(rec("u1", "t1")).join();
        assertEquals(0, sink.purge("nobody").join());
    }
}
